### PR TITLE
feat: add pnpm store visualization command

### LINKED ISSUE #7413: Enhance Visualization of .pnpm-store
### Contribution

- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

I have been using pnpm and find it really useful, especially the way it optimizes package storage using its store. However, I find it really hard sometimes to understand where packages are placed within the .pnpm-store due to its non-human-readable, hash-based structure.

### Describe the solution you'd like

I think it would be a huge benefit if there was a way, maybe a command or tool, that can provide a visualization or mapping from package names (plus versions) to their locations inside .pnpm-store.

This could help with various debugging scenarios, and generally make the location of packages more transparent to the user.

Does something like this exist, or could it be a new feature in a future version of pnpm? I would love to hear your thoughts on this.

Thank you very much!

### Describe the drawbacks of your solution

-

### Describe alternatives you've considered

-

### PR DESCRIPTION
**Pull request by byteHulk**
~Might need to make up some unit tests as well~'

close #7413

### CODE REVIEW

File: store/plugin-commands-store-search/src/catFile.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+export function help () {
+  return renderHelp({
+    description:
+      'Prints the contents of a file based on the hash value stored in the index file.',
+    descriptionLists: [],
+    usages: ['pnpm cat-file <hash>'],
+  })
+}
+
+export type catFileCommandOptions = Pick<Config, 'storeDir' | 'pnpmHomeDir'>
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catFile.ts]:
```suggestion
export type CatFileCommandOptions = Pick<Config, 'storeDir' | 'pnpmHomeDir'>
```

File: store/plugin-commands-store-search/src/catFile.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+export async function handler (opts: catFileCommandOptions, params: string[]) {
+  if (!params || params.length === 0) {
+    throw new PnpmError('MISSING_HASH', '`pnpm cat-file` requires the hash')
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catFile.ts]:
```suggestion
    throw new PnpmError('MISSING_HASH', 'Missing file hash', {
      hint: help(),
    })
```

File: store/plugin-commands-store-search/src/catFile.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  const storeDir = await getStorePath({
+    pkgRoot: process.cwd(),
+    storePath: opts.storeDir,
+    pnpmHomeDir: opts.pnpmHomeDir,
+  })
+  const cafsDir = path.join(storeDir, 'files')
+  const filePath = path.resolve(cafsDir, toHex.slice(0, 2), toHex.slice(2))
+
+  try {
+    return await gfs.readFile(filePath, 'utf8')
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catFile.ts]: This should be sync

File: store/plugin-commands-store-search/src/catFile.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  } catch {
+    throw new PnpmError('INVALID_HASH', 'Corresponding hash file not found')
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catFile.ts]: This error makes sense only on ENOENT. Other errors should pass through.

File: store/plugin-commands-store-search/src/catIndex.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+| 'registries'
+| 'cacheDir'
+>
+
+export async function handler (opts: catIndexCommandOptions, params: string[]) {
+  if (!params || params.length === 0) {
+    throw new PnpmError(
+      'MISSING_PACKAGE_NAME',
+      '`pnpm cat-index` requires the package name'
+    )
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catIndex.ts]:
```suggestion
    throw new PnpmError(
      'MISSING_PACKAGE_NAME',
      'Specify a package',
      {
        hint: help(),
      }
    )
```

File: store/plugin-commands-store-search/src/catIndex.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    (pkgSnapshot.resolution as TarballResolution).integrity!.toString(),
+    'index'
+  )
+  try {
+    const pkgFilesIndex = await loadJsonFile<PackageFilesIndex>(filesIndexFile)
+    return JSON.stringify(pkgFilesIndex, null, 2)
+  } catch {
+    throw new PnpmError(
+      'INVALID_PACKAGE',
+      'No corresponding index file found. You can use pnpm list to see if the package is installed.'
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catIndex.ts]: This error only makes sense on ENOENT. Also, the try/catch should only cover the line with loadJsonFile

File: store/plugin-commands-store-search/src/catIndex.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+export function help () {
+  return renderHelp({
+    description:
+      'Prints the corresponding index file based on package name and version.',
+    descriptionLists: [],
+    usages: ['pnpm cat-index <pkg name>@<pkg version>'],
+  })
+}
+
+export type catIndexCommandOptions = Pick<
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/catIndex.ts]:
```suggestion
export type CatIndexCommandOptions = Pick<
```

File: store/plugin-commands-store-search/src/findHash.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  return renderHelp({
+    description:
+      'Print the contained packages according to the hash.',
+    descriptionLists: [],
+    usages: ['pnpm find-hash <hash>'],
+  })
+}
+
+export type findHashCommandOptions = Pick<Config, 'storeDir' | 'pnpmHomeDir'>
+export interface findHashResult {
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/findHash.ts]:
```suggestion
export type FindHashCommandOptions = Pick<Config, 'storeDir' | 'pnpmHomeDir'>
export interface FindHashResult {
```

File: store/plugin-commands-store-search/src/findHash.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  const hash = params[0]
+  const storeDir = await getStorePath({
+    pkgRoot: process.cwd(),
+    storePath: opts.storeDir,
+    pnpmHomeDir: opts.pnpmHomeDir,
+  })
+  const cafsDir = path.join(storeDir, 'files')
+  const cafsChildrenDirs = fs.readdirSync(cafsDir).filter(dirName => fs.statSync(`${cafsDir}/${dirName}`).isDirectory())
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/findHash.ts]: Use readdirSync with the withFileTypes option set to true. In that case you won't have to run statSync.

File: store/plugin-commands-store-search/src/findHash.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    const dirIndexFiles = fs
+      .readdirSync(`${cafsDir}/${dirName}`)
+      .filter((fileName) => fileName.includes('-index.json'))
+      ?.map((fileName) => `${cafsDir}/${dirName}/${fileName}`)
+
+    indexFiles.push(...dirIndexFiles)
+  })
+
+  indexFiles.forEach(indexPath => {
+    const data = JSON.parse(fs.readFileSync(indexPath, 'utf-8') || '{}')
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/findHash.ts]: This a bit confusing. If you really want to handle the case when the index file is empty, the do
```suggestion
    const content = fs.readFileSync(indexPath, 'utf-8')
    if (!content) return
    const data = JSON.parse(content)
```

File: store/plugin-commands-store-search/src/findHash.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    if (!data?.name || !data?.files) return
+    Object.keys(data.files).forEach(key => {
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/findHash.ts]: the file can be in the sideEffects field to.

File: store/plugin-commands-store-search/src/findHash.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      if (data.files?.[key]?.integrity === hash) {
+        result.push({ name: data.name, version: data?.version || 'latest', indexPath: indexPath.replace(cafsDir, '') })
```
Comment (zkochan) [inline store/plugin-commands-store-search/src/findHash.ts]: it doesn't make sense to set it to latest. If you want to set it to something, I guess `unknown` could be an option.

**Review by zkochan — CHANGES_REQUESTED** [review]

File: store/plugin-commands-store-search/src/catFile.ts (hunk 1 above)
Comment (byteHulk) [reply store/plugin-commands-store-search/src/catFile.ts]: Thank you. I'll check each one.

File: store/plugin-commands-store-search/src/catFile.ts (hunk 3 above)
Comment (byteHulk) [reply store/plugin-commands-store-search/src/catFile.ts]: May I ask what is the reason for using synchronization here?

This helps me contribute code more efficiently😂

Comment (byteHulk) [reply store/plugin-commands-store-search/src/catFile.ts]: I'll make the changes first.

**Comment by byteHulk** [comment]
All changes have been implemented

File: store/plugin-commands-store-search/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# @pnpm/plugin-commands-store-search
+
+> The visualization store commands of pnpm
+
+[![npm version](https://img.shields.io/npm/v/@pnpm/plugin-commands-store-search.svg)](https://www.npmjs.com/package/@pnpm/plugin-commands-store-search)
+
+## Installation
+
+```sh
+pnpm add @pnpm/plugin-store-search
```
Comment (zkochan) [inline store/plugin-commands-store-search/README.md]: This doesn't match the package name.

File: store/plugin-commands-store-search/package.json [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    "load-json-file": "^6.2.0",
+    "tempy": "^1.0.1",
+    "execa": "npm:safe-execa@0.1.2"
+  },
+  "dependencies": {
+    "@pnpm/error": "workspace:*",
+    "@pnpm/types": "workspace:*",
+    "render-help": "^1.0.3",
+    "chalk": "^4.1.2"
+  },
```
Comment (zkochan) [inline store/plugin-commands-store-search/package.json]: These dependencies are broken. Some of the dev dependencies are used in runtime code, they should be moved to prod dependencies.

File: store/plugin-commands-store-search/README.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
+# @pnpm/plugin-commands-store-search
```
Comment (zkochan) [inline store/plugin-commands-store-search/README.md]: I think `plugin-commands-store-inspecting` will be a better name for this package.

File: store/plugin-commands-store-search/package.json [hunk 16]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@pnpm/plugin-commands-store-search",
+  "version": "0.0.1",
```
Comment (zkochan) [inline store/plugin-commands-store-search/package.json]: Start with 0.0.0

File: store/plugin-commands-store-search/README.md (hunk 15 above)
Comment (byteHulk) [reply store/plugin-commands-store-search/README.md]: okay

**Comment by zkochan** [comment]
add a changeset.

**Comment by byteHulk** [comment]
The official website documentation also needs to be updated?

**Comment by zkochan** [comment]
yes, when it gets released.

**Comment by byteHulk** [comment]
Let me try?

**Comment by zkochan** [comment]
The docs are here: https://github.com/pnpm/pnpm.io

But first this PR should be finished.

**Comment by byteHulk** [comment]
Is there anything else I can do?

**Comment by zkochan** [comment]
I added a comment to the issue. We need to discuss that. https://github.com/pnpm/pnpm/issues/7413#issuecomment-1868264468

**Comment by zkochan** [comment]
In the output of find-hash it might make more sense to print the integrity of the package instead of the relative path to the index.json file. Or if the path is printed out, then the whole path should be printed, including the store directory.

**Comment by byteHulk** [comment]
like this?

`lodash@4.17.19  /Users/pnpm/store/v3/files/24/dbddf17111f46417d2fdaa260b1a37f9b3142340e4145efe3f0937d77eb56c862d2a1d2901ca16271dc0d6335b0237c2346768a3ec1a3d579018f1fc5f7a0d-index.json`

I do think it's a little long, it's all duplicate paths, it's distracting to the user.

**Review by zkochan — APPROVED** [review]

**Pull request merged**